File: app/src/english/java/com/nagopy/android/textcounter/model/EnglishTextCounter.java
package com.nagopy.android.textcounter.model;

import android.support.annotation.NonNull;

import java.util.regex.Pattern;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class EnglishTextCounter extends TextCounter {

    Pattern regexSpaces = Pattern.compile("\\s");
    Pattern regexMultiBr = Pattern.compile("\n{2,}");
    Pattern regex = Pattern.compile("[\\.!\\?\\s]+");

    @Inject
    public EnglishTextCounter() {
    }

    @Override
    int countChars(@NonNull String text) {
        return remove(text, regexSpaces).length();
    }

    @Override
    int countWords(@NonNull String text) {
        String wkStr = text.replace('\n', '.');
        wkStr = remove(wkStr, "-");
        wkStr = replace(wkStr, regex, ".");

        int result = count(wkStr, ".");

        if (wkStr.startsWith(".")) {
            result--;
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith(".")) {
            result++;
        }

        return result;
    }

    @Override
    int countSentences(@NonNull String text) {
        String wkStr = remove(text, regexSpaces);
        wkStr = replace(wkStr, regex, "."); // 連続しているものは全て一つのピリオドに

        // ピリオドを数えることで、センテンス数を数える
        int result = count(wkStr, ".");

        // 補正
        if (wkStr.startsWith(".")) {
            result--;
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith(".")) {
            result++;
        }

        return result;
    }

    @Override
    int countParagraphs(@NonNull String text) {
        String wkStr = text;
/*
        // 文字列が改行のみの場合はカウントを中止
        String wkStr = remove(text, regexSpaces);
        if (wkStr.isEmpty()) {
            return 0;
        }
*/

        // 連続した改行は一つにまとめる
        wkStr = replace(wkStr, regexMultiBr, "\n");

        // 改行をカウント
        int result = count(wkStr, "\n");

        // 補正
        if (wkStr.startsWith("\n")) {
            result--;
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith("\n")) {
            result++;
        }

        return result;
    }
}

File: app/src/main/java/com/nagopy/android/textcounter/counter/EnglishTextCounter.kt
/*
 * Copyright 2018 75py
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nagopy.android.textcounter.counter

class EnglishTextCounter : TextCounter {

    val regexSpaces = "\\s".toRegex()
    val regexHyphen = "-".toRegex()
    val regexDelimiter = "[.!?\\s]+".toRegex()
    val regexPeriod = "\\.".toRegex()
    val regexMultiBr = "\n{2,}".toRegex()
    val regexBr = "\n".toRegex()
    val regexWhitespace = " ".toRegex()

    override fun countChars(inputText: String): Int {
        return inputText.remove(regexSpaces).length
    }

    override fun countWords(inputText: String): Int {
        var wkStr = inputText.replace('\n', '.')
        wkStr = wkStr.remove(regexHyphen)
        wkStr = wkStr.replace(regexDelimiter, ".")

        var result = wkStr.count(regexPeriod)

        if (wkStr.startsWith(".")) {
            result--
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith(".")) {
            result++
        }

        return result
    }

    override fun countSentences(inputText: String): Int {
        var wkStr = inputText.remove(regexSpaces)
        wkStr = wkStr.replace(regexDelimiter, ".") // 連続しているものは全て一つのピリオドに

        // ピリオドを数えることで、センテンス数を数える
        var result = wkStr.count(regexPeriod)

        if (wkStr.startsWith(".")) {
            result--
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith(".")) {
            result++
        }

        return result
    }

    override fun countParagraphs(inputText: String): Int {
        var wkStr = inputText

        // 連続した改行は一つにまとめる
        wkStr = wkStr.replace(regexMultiBr, "\n")

        // 改行をカウント
        var result = wkStr.count(regexBr)

        // 補正
        if (wkStr.startsWith("\n")) {
            result--
        }

        if (!wkStr.isEmpty() && !wkStr.endsWith("\n")) {
            result++
        }

        return result
    }

    override fun countWhitespaces(inputText: String): Int {
        return inputText.count(regexWhitespace)
    }

}